File: src/local_harness/application/timeline.py
"""Pure request-level activity timeline projection."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from local_harness.domain.models import ProgressEvent


@dataclass(frozen=True, slots=True)
class TimelineStep:
    """One concise observable step projected from persisted progress events."""

    key: str
    label: str
    target: str
    status: Literal["started", "success", "warning", "error"]
    duration_ms: int
    sequence: int


def build_request_timeline(
    events: list[ProgressEvent], request_number: int | None
) -> list[TimelineStep]:
    """Merge each model tool request with its matching tool result."""
    relevant = [event for event in events if event.request_number == request_number]
    steps: list[TimelineStep] = []
    consumed_tools: set[int] = set()
    for event in relevant:
        if event.kind == "model_start":
            continue
        if event.kind == "model_complete" and event.target != "final":
            tools = [
                candidate
                for candidate in relevant
                if candidate.call_number == event.call_number
                and candidate.kind in {"tool_complete", "tool_error", "plan_update"}
                and candidate.sequence not in consumed_tools
            ]
            if tools:
                for tool in tools:
                    consumed_tools.add(tool.sequence)
                    steps.append(
                        TimelineStep(
                            f"tool-{tool.sequence}",
                            event.summary,
                            tool.target,
                            tool.status,
                            event.duration_ms + tool.duration_ms,
                            tool.sequence,
                        )
                    )
                continue
        if event.sequence in consumed_tools or event.kind in {
            "tool_complete",
            "tool_error",
            "plan_update",
        }:
            if event.sequence in consumed_tools:
                continue
        steps.append(
            TimelineStep(
                f"event-{event.sequence}",
                event.summary,
                event.target,
                event.status,
                event.duration_ms,
                event.sequence,
            )
        )
    return steps

File: web/src/timeline.ts
import type { ProgressEvent } from "./types";

export interface TimelineStep {
  key: string;
  label: string;
  target: string;
  status: ProgressEvent["status"];
  duration_ms: number;
  sequence: number;
}

export function buildRequestTimeline(
  events: ProgressEvent[],
  requestNumber: number | null,
): TimelineStep[] {
  const relevant = events.filter(
    (event) => event.request_number === requestNumber,
  );
  const consumed = new Set<number>();
  const steps: TimelineStep[] = [];
  for (const event of relevant) {
    if (event.kind === "model_start") continue;
    if (event.kind === "model_complete" && event.target !== "final") {
      const tools = relevant.filter(
        (candidate) =>
          candidate.call_number === event.call_number &&
          ["tool_complete", "tool_error", "plan_update"].includes(
            candidate.kind,
          ) &&
          !consumed.has(candidate.sequence),
      );
      if (tools.length) {
        for (const tool of tools) {
          consumed.add(tool.sequence);
          steps.push({
            key: `tool-${tool.sequence}`,
            label: event.summary,
            target: tool.target,
            status: tool.status,
            duration_ms: event.duration_ms + tool.duration_ms,
            sequence: tool.sequence,
          });
        }
        continue;
      }
    }
    if (consumed.has(event.sequence)) continue;
    steps.push({
      key: `event-${event.sequence}`,
      label: event.summary,
      target: event.target,
      status: event.status,
      duration_ms: event.duration_ms,
      sequence: event.sequence,
    });
  }
  return steps;
}
